model_zoo/lbann.cpp

```diff
@@ -43,6 +43,10 @@ int main(int argc, char *argv[]) {
   world_comm_ptr comm = initialize(argc, argv, random_seed);
   const bool master = comm->am_world_master();
 
+  MPI_Errhandler err_handle;
+  MPI_Comm_create_errhandler(lbann_mpi_err_handler, &err_handle);
```

Comment:
Do we also need to deallocate the error handler with [`MPI_Errhandler_free`](https://www.open-mpi.org/doc/current/man3/MPI_Errhandler_free.3.php)? The documentation for [`MPI_Comm_create_errhandler`](https://www.open-mpi.org/doc/current/man3/MPI_Comm_create_errhandler.3.php) doesn't mention it and quick Googling doesn't find any guidelines. If we go that route, I think it would be nicer to put this inside of `lbann::initialize`/`lbann::finalize` or to wrap the `MPI_Errhandler` inside of a RAII object.

Reply:
If you want to live in a shiny world of RAII purity, you should free the resources you acquire.

In this case, `MPI_Errhandler_free` will do nothing useful. It won't deallocate the error handler until `MPI_COMM_WORLD` is deallocated, which won't happen until the end of the program (anything after `MPI_Finalize()` is undefined behavior in an MPI program). All program resources are freed and reclaimed by the OS at the end of the program, which will include this error handler. Depending on the MPI implementation you're using, you _might_ see a bit of "leaked" memory in a tool like LeakSanitizer or Valgrind, but this could be (should be) ignored.